### PR TITLE
Added gps and mag UAVCAN support to SITL

### PR DESCRIPTION
**Pull request by d-v**
Added gps and mag over UAVCAN simulation to SITL. Tested on bare metal ubuntu 16.04 LTS.

Added another launch arg to sim_vehicle.py: --uavcan <interface>. I tested on vcan0 under ArduPilot. When UAVCAN is enabled, sim_vehicle.py uses <vehicle>-uavcan.parm. I created copter-uavcan.parm but haven't done so for other vehicles. copter-uavcan.parm is essentially the same as copter.parm, just some uavcan-specific params used.

sim_vehicle.py --console --uavcan vcan0 -j4 is what I used to run SITL.

Many thanks to everyone who answered my questions on Gitter and Skype, I am really grateful for the help. Looking forward to hearing everyone's feedback!

### CODE REVIEW

**Comment by khancyr** [comment]
@d-v To help us on reviewing, start by separate your commit by functionnality ! 
http://ardupilot.org/dev/docs/git-interactive-rebase.html and the script git-subsystems-split in Tools/gittools will help you doing that !

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -307,7 +329,7 @@ def configure_env(self, cfg, env):
             '-Wno-error=undef',
             '-Wno-error=cpp',
             '-Wno-cast-align',
-            '-fno-exceptions',
+            '-fexceptions',
```
Comment (khancyr) [inline Tools/ardupilotwaf/boards.py]: remove that

File: Tools/ardupilotwaf/boards.py [hunk 2]
```diff
@@ -261,12 +269,26 @@ def configure_env(self, cfg, env):
...
+
         if sys.platform == 'cygwin':
             env.LIB += [
                 'winmm',
             ]
             env.CXXFLAGS += ['-DCYGWIN_BUILD']
 
+        if self.with_uavcan:
+            env.DEFINES.update(
+                UAVCAN_CPP_VERSION = 'UAVCAN_CPP11'
```
Comment (khancyr) [inline Tools/ardupilotwaf/boards.py]: maybe just separate this one on another pr. It is should only update the board config

File: Tools/autotest/default_params/copter-uavcan.parm [hunk 3]
```diff
@@ -0,0 +1,97 @@
+CAN_P1_DRIVER   1.000000
```
Comment (khancyr) [inline Tools/autotest/default_params/copter-uavcan.parm]: Look at how it is done on other vehicle type and config.
We should simply append only the config for uavcan not redefine all param ...

File: Tools/autotest/default_params/copter.parm [hunk 4]
```diff
@@ -54,6 +54,7 @@ SIM_MAG_RND     0
 SIM_GPS_GLITCH_X    0
 SIM_GPS_GLITCH_Y    0
 SIM_GPS_GLITCH_Z    0
+GPS_TYPE         1.000000
```
Comment (khancyr) [inline Tools/autotest/default_params/copter.parm]: this is already the case no?

File: Tools/autotest/sim_vehicle.py [hunk 5]
```diff
@@ -566,6 +566,9 @@ def start_vehicle(binary, autotest, opts, stuff, loc):
     path = None
     if "default_params_filename" in stuff:
         paths = stuff["default_params_filename"]
+        if opts.uavcan:
+            suffix = paths.find(".")
+            paths = paths[:suffix]+"-uavcan"+paths[suffix:]
```
Comment (khancyr) [inline Tools/autotest/sim_vehicle.py]: paths = paths[:suffix] + "-uavcan" + paths[suffix:]

Comment (khancyr) [reply Tools/autotest/sim_vehicle.py]: see my previos comment, it will be simplier to load only the config for uavacan instead of all params

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 6]
```diff
@@ -80,15 +101,26 @@ void SITL_State::_sitl_setup(const char *home_str)
     _barometer = (AP_Baro *)AP_Param::find_object("GND_");
     _ins = (AP_InertialSensor *)AP_Param::find_object("INS_");
     _compass = (Compass *)AP_Param::find_object("COMPASS_");
+#if HAL_WITH_UAVCAN
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (khancyr) [inline libraries/AP_HAL_SITL/SITL_State.cpp]: not needed as you are already on SITL namespace

File: Tools/ardupilotwaf/boards.py [hunk 7]
```diff
@@ -261,12 +269,26 @@ def configure_env(self, cfg, env):
             'SITL',
         ]
 
+        env.GIT_SUBMODULES += [
+            'uavcan',
```
Comment (khancyr) [inline Tools/ardupilotwaf/boards.py]: I think it should check for SLCAN or whatever is the name for the linux implementation of CAN bus !

File: libraries/AP_Param/AP_Param.h [hunk 8]
```diff
@@ -402,7 +402,14 @@ class AP_Param
     bool configured_in_storage(void) const;
 
     // return true if the parameter is configured
-    bool configured(void) const { return configured_in_defaults_file() || configured_in_storage(); }
+    bool configured(void) const
+    {
+#if HAL_WITH_UAVCAN
+        return true;
```
Comment (khancyr) [inline libraries/AP_Param/AP_Param.h]: this doesn't look good

**Comment by d-v** [comment]
@khancyr, thank you for all of the feedback it was really helpful especially on my first real PR. I'll incorporate these changes and update the PR.

File: Tools/autotest/sim_vehicle.py (hunk 5 above)
Comment (d-v) [reply Tools/autotest/sim_vehicle.py]: @khancyr, I'm sorry I don't understand what you mean :(

File: Tools/ardupilotwaf/boards.py [hunk 9]
```diff
@@ -110,7 +110,7 @@ def configure_env(self, cfg, env):
 
             '-fdata-sections',
             '-ffunction-sections',
-            '-fno-exceptions',
+            '-fexceptions',
```
[the remark this answers is not in the document]
Comment (d-v) [reply Tools/ardupilotwaf/boards.py]: It didn't work :(
https://gist.github.com/d-v/508d4c66f548f3301f0384dd627828bf
Even with cfg.define(UAVCAN_EXCEPTIONS, 0)

**Comment by magicrub** [comment]
Please take a look at the guidelines
http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

commit comments should start with the directory name.

Also, you now have files in this PR, so something went crazy again. I'm sure a rebase would fix it

**Comment by d-v** [comment]
@magicrub, thanks for the feedback -- sorry that this has been such a messy PR! I'm still working on one residual issue that I'm fairly certain will allow all the tests to pass. Once that's done I'll be able to clean up and break this up into a bunch of commits that are properly styled. Sorry about the confusion!

**Comment by davidbuzz** [comment]
i liked this old PR enough to try re-working it based on current ArduPilot.  My rework of the code doesn't seem to compile right now when I add  "--uavcan vcan0" to sim_vehicle.py , as a result of  libuavcan incompatibilities with the linux can driver etc.    anyway.... i'm wondering if the original author of this PR, or anyone else, is interested enough to help me figure out where my revised branch needs work...   https://github.com/davidbuzz/ardupilot/tree/uavcan-sitl-buzz-wip  ?

**Comment by khancyr** [comment]
We got a proper implementation using AP_Periph for those now. I am closing this. Thanks anyway
